Reference Architectures (#3712)
fleetdm/fleet: docs/02-Deploying/06-Reference-Architectures.md

# Reference Architectures

You can easily run Fleet on a single VPS that would be capable of supporting hundreds if not thousands of hosts, but

> this page will detail what an [opinionated view](https://github.com/fleetdm/fleet/tree/main/tools/terraform) of running Fleet in a production environment, as

Reviewer: ```suggestion
this page details an [opinionated view](https://github.com/fleetdm/fleet/tree/main/tools/terraform) of running Fleet in a production environment, as
```

well as different configuration strategies to enable High Availability (HA).

## Availability Components

There are a few strategies that can be used to ensure high availability:
- Database HA
- Traffic load balancing

### Database HA

Fleet recommends RDS Aurora MySQL when running on AWS. More details about backups/snapshots can be found
[here](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Aurora.Managing.Backups.html). It is also
possible to dynamically scale read replicas to increase performance and [enable database fail-over](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Concepts.AuroraHighAvailability.html).
It is also possible to use [Aurora Global](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-global-database.html) to
span multiple regions for more advanced configurations(_not included in the [reference terraform](https://github.com/fleetdm/fleet/tree/main/tools/terraform)_).

In some cases adding a read replica can increase database performance for specific access patterns. In scenarios when automating the API or with `fleetctl`

> there can be benefits to read performance. Ultimately

Reviewer: ```suggestion
there can be benefits to read performance.
```

### Traffic load balancing
Load balancing enables distributing request traffic over many instances of the backend application. Using AWS Application
Load Balancer can also [offload SSL termination](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/create-https-listener.html), freeing Fleet to spend the majority of it's allocated compute dedicated 
to its core functionality. More details about ALB can be found [here](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html).

_**Note if using [terraform reference architecture](https://github.com/fleetdm/fleet/tree/main/tools/terraform#terraform) all configurations can dynamically scale based on load(cpu/memory) and all configurations

> assume On-Demand pricing. Calculations do not take into account NAT gateway charges or other networking related ingress/egress costs.**_

Reviewer: ```suggestion
assume On-Demand pricing (savings are available through Reserved Instances). Calculations do not take into account NAT gateway charges or other networking related ingress/egress costs.**_
```

|-----------------|---------------|-----|
| 10 Fargate task  | 1024 CPU Units | 4GB |

| Dependencies | Version                 | Instance type |
|--------------|-------------------------|---------------|
| Redis        | 6                       |  m6g.large    |
| MySQL        | 5.7.mysql_aurora.2.10.0 | db.r6g.large  |

#### [Up to 150000 hosts](https://calculator.aws/#/estimate?id=6a852ef873c0902f0c953045dec3e29fcd32aef8)

| Fleet instances | CPU Units      | RAM |
|-----------------|----------------|-----|
| 30 Fargate task | 1024 CPU Units | 4GB |

| Dependencies | Version                 | Instance type  | Nodes |
|--------------|-------------------------|----------------|-------|
| Redis        | 6                       | m6g.large      | 3     |
| MySQL        | 5.7.mysql_aurora.2.10.0 | db.m6g.8xlarge | 1     |

## Cloud Providers

### AWS

AWS reference architecture can be found [here](https://github.com/fleetdm/fleet/tree/main/tools/terraform). This configuration includes:

- VPC
  - Subnets
    - Public & Private
  - ACLs
  - Security Groups
- ECS as the container orchestrator
  - Fargate for underlying compute
  - Task roles via IAM
- RDS Aurora MySQL 5.7

> - Elasicache Redis Engine

Reviewer: ```suggestion
- Elasticache Redis Engine
```

- Firehose osquery log destination
  - S3 bucket sync to allow further ingestion/processing
- [Monitoring via Cloudwatch alarms](https://github.com/fleetdm/fleet/tree/main/tools/terraform/monitoring)

Some AWS services used in the provider reference architecture are billed as pay-per-use such as Firehose. This means that osquery scheduled query frequency can have

> a direct correlation to how much these services cost, something to at least keep in mind when configuring Fleet in AWS while using some logging destinations.

Reviewer: ```suggestion
a direct correlation to how much these services cost, something to keep in mind when configuring Fleet in AWS.
```
